aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -613,6 +606,43 @@ Tensor& linalg_matrix_rank_out(
   return result;
 }
 
+Tensor get_matrix_rank_result_tensor(const Tensor& input) {
+  // Matrices or batch of matrices are allowed
+  TORCH_CHECK(
+      input.dim() >= 2,
+      "torch.linalg.matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ",
+      input.sizes());
+  // For Composite Compliance, allocate `result` of correct shape to
+  // avoid resizing in `out` variant.
+  // See also `NOTE [matrix rank output shape]`
+  auto result_shape =
+      IntArrayRef(input.sizes().cbegin(), input.sizes().cend() - 2);
+  Tensor result =
+      at::empty(result_shape, input.options().dtype(ScalarType::Long));
+}
+
+}  // anonymous namespace
+
+Tensor& linalg_matrix_rank_out(
+    const Tensor& input,
+    const optional<Tensor>& atol_opt,
+    const optional<Tensor>& rtol_opt,
+    bool hermitian,
+    Tensor& result) {
+  // Matrices or batch of matrices are allowed
+  TORCH_CHECK(
+      input.dim() >= 2,
+      "torch.linalg.matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ",
+      input.sizes());
+  // NOTE [matrix rank output shape]
+  // matrix_rank assigns a scalar value for each matrix in the batch so
+  // result's shape is equal to input.shape[0:input.ndim-2]
+  // for single matrix result_shape = {}
+  auto result_shape = IntArrayRef(input.sizes().cbegin(), input.sizes().cend() - 2);
+  at::native::resize_output(result, result_shape);
```

Comment:
Skipping this check here and resize conditioned on `input.dim` will not work?

Reply:
But we do want the check if `out` variant is called directly from Python, right?

Reply:
But the `impl` is doing all the check, or no?

Reply:
Have moved this particular check out of `impl`. 

All the functional variants do the check upfront via `get_matrix_rank_result_tensor` and `out` variant does it here since it can be called from Python with incorrect shape

Reply:
All right, got it! Still these lines are identical to what `get_matrix_rank_result_tensor` is doing.